tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/ValidateAMQPErrorsTest.java

```diff
@@ -0,0 +1,600 @@
+      }
+   }
+
+   @Test
+   public void testReconnectAfterSenderOpenTimeout() throws Exception {
+
+      startVerx();
+
+      ProtonServerOptions serverOptions = new ProtonServerOptions();
+
+      AtomicInteger countOpen = new AtomicInteger(0);
+      CyclicBarrier startFlag = new CyclicBarrier(2);
+      CountDownLatch blockBeforeOpen = new CountDownLatch(1);
+      AtomicInteger disconnects = new AtomicInteger(0);
+      AtomicInteger messagesReceived = new AtomicInteger(0);
+
+      ConcurrentHashSet<ProtonConnection> connections = new ConcurrentHashSet<>();
+
+      mockServer = new MockServer(vertx, serverOptions, null, serverConnection -> {
+         serverConnection.disconnectHandler(c -> {
+            disconnects.incrementAndGet(); // number of retries
+            connections.remove(c);
+         });
+         serverConnection.openHandler(serverSender -> {
+            serverConnection.closeHandler(x -> {
+               serverConnection.close();
+               connections.remove(serverConnection);
+            });
+            serverConnection.open();
+            connections.add(serverConnection);
+         });
+         serverConnection.sessionOpenHandler((s) -> {
+            s.open();
+         });
+         serverConnection.senderOpenHandler((x) -> {
+            x.open();
+         });
+         serverConnection.receiverOpenHandler((x) -> {
+            if (countOpen.incrementAndGet() > 2) {
+               if (countOpen.get() == 3) {
+                  try {
+                     startFlag.await(10, TimeUnit.SECONDS);
+                     blockBeforeOpen.await(10, TimeUnit.SECONDS);
+                     return;
+                  } catch (Throwable ignored) {
+                  }
+               }
+               HashMap<Symbol, Object> brokerIDProperties = new HashMap<>();
+               brokerIDProperties.put(AMQPMirrorControllerSource.BROKER_ID, "fake-id");
+               x.setProperties(brokerIDProperties);
+               x.setOfferedCapabilities(new Symbol[]{AMQPMirrorControllerSource.MIRROR_CAPABILITY});
+               x.setTarget(x.getRemoteTarget());
+               x.open();
+               x.handler((del, msg) -> {
+                  System.out.println("prefetch = " + x.getPrefetch() + ", Got message: " + msg);
+                  if (msg.getApplicationProperties() != null) {
+                     Map map = msg.getApplicationProperties().getValue();
+                     Object value = map.get("sender");
+                     if (value != null && value.equals("jms")) {
+                        messagesReceived.incrementAndGet();
+                     }
+                  } else {
+                     System.out.println("app properties == null");
+                  }
+                  del.settle();
+                  if (x.getPrefetch() == 0) {
+                     x.flow(1);
+                  }
+               });
+            }
+         });
+      });
+
+      AssertionLoggerHandler.startCapture();
```

Comment:
Oops, it was this one I meant to comment on before that it seemed unused.

Reply:
that one was actually used.. thanks.. I will remove it..

I was using it at some point during initial dev, then I replaced it by the messagesReceived value.